### ISSUE TITLE
Read access violation crash when closing game if a node is no longer a child of the scene tree.

### ISSUE THREAD

**Issue by JonathaN7Shepard** [godotengine/godot#84265]
### Godot version

v4.1.2.stable.mono.official [399c9dc39], v4.2.beta.mono.custom_build [93cdacbb0]

### System information

Godot v4.1.2.stable.mono - Windows 10.0.22621 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 3070 (NVIDIA; 31.0.15.3742) - AMD Ryzen 9 5950X 16-Core Processor (32 Threads)

### Issue description

When a node is tweened and then removed from its parent, the engine crashes with a read access violation when closing the game. I never call queue_free() on the node in question. I realize this node is dangling, but the engine should not crash on close because of it.

The tween guarantees the issue can be reproduced. When closing the game, an access violation happens here (the singleton must have been deleted in cleanup before reaching this code where it's dereferenced):
https://github.com/godotengine/godot/blob/4363ae34fa3f52223ae78ff2f2c187ab99827e3a/servers/rendering/renderer_canvas_render.h#L154

I've tried it without the tween in different variations, and sometimes I can get a crash in the DEV_ASSERT here:
https://github.com/godotengine/godot/blob/4363ae34fa3f52223ae78ff2f2c187ab99827e3a/core/templates/self_list.h#L165

I have not been able to reproduce the issue in GDScript, only C#.

### Steps to reproduce

Open repro project, run it, press spacebar, close game, engine crashes on close.

### Minimal reproduction project

[TweenAccessViolation.zip](https://github.com/godotengine/godot/files/13221230/TweenAccessViolation.zip)

**Comment by raulsntos** [comment]
> I have not been able to reproduce the issue in GDScript, only C#.

How did you test this? I'm able to reproduce the crash in GDScript too. Here's the ported project:

[TweenAccessViolationGD.zip](https://github.com/godotengine/godot/files/13229667/TweenAccessViolationGD.zip)

Since it crashes when freeing the Rendering Server, it seems like a rendering bug. I've been able to avoid the crash with these changes, but I'm not sure if it's the correct fix:

```diff
diff --git a/servers/rendering/renderer_canvas_render.h b/servers/rendering/renderer_canvas_render.h
index ef4de9ce54..4ac4d753e4 100644
--- a/servers/rendering/renderer_canvas_render.h
+++ b/servers/rendering/renderer_canvas_render.h
@@ -150,7 +150,7 @@ public:
 
 		_FORCE_INLINE_ Polygon() { polygon_id = 0; }
 		_FORCE_INLINE_ ~Polygon() {
-			if (polygon_id) {
+			if (singleton && polygon_id) {
 				singleton->free_polygon(polygon_id);
 			}
 		}
@@ -523,7 +523,9 @@ public:
 	virtual void set_debug_redraw(bool p_enabled, double p_time, const Color &p_color) = 0;
 
 	RendererCanvasRender() { singleton = this; }
-	virtual ~RendererCanvasRender() {}
+	virtual ~RendererCanvasRender() {
+		singleton = nullptr;
+	}
 };
 
 #endif // RENDERER_CANVAS_RENDER_H

```

**Comment by JonathaN7Shepard** [comment]
> > I have not been able to reproduce the issue in GDScript, only C#.
> 
> How did you test this? I'm able to reproduce the crash in GDScript too. Here's the ported project:
> 
> [TweenAccessViolationGD.zip](https://github.com/godotengine/godot/files/13229667/TweenAccessViolationGD.zip)
> 
> Since it crashes when freeing the Rendering Server, it seems like a rendering bug. I've been able to avoid the crash with these changes, but I'm not sure if it's the correct fix:
> 
> ```diff
> diff --git a/servers/rendering/renderer_canvas_render.h b/servers/rendering/renderer_canvas_render.h
> index ef4de9ce54..4ac4d753e4 100644
> --- a/servers/rendering/renderer_canvas_render.h
> +++ b/servers/rendering/renderer_canvas_render.h
> @@ -150,7 +150,7 @@ public:
>  
>  		_FORCE_INLINE_ Polygon() { polygon_id = 0; }
>  		_FORCE_INLINE_ ~Polygon() {
> -			if (polygon_id) {
> +			if (singleton && polygon_id) {
>  				singleton->free_polygon(polygon_id);
>  			}
>  		}
> @@ -523,7 +523,9 @@ public:
>  	virtual void set_debug_redraw(bool p_enabled, double p_time, const Color &p_color) = 0;
>  
>  	RendererCanvasRender() { singleton = this; }
> -	virtual ~RendererCanvasRender() {}
> +	virtual ~RendererCanvasRender() {
> +		singleton = nullptr;
> +	}
>  };
>  
>  #endif // RENDERER_CANVAS_RENDER_H
> ```

You're right, I can reproduce the crash in GDScript too. My testing method was flawed as I wasn't running the engine in the debugger. I realized that setting the working directory to the project will open the game up for debugging. I'm seeing the crash now.

I'm not sure if that's the correct fix either, I've been using Godot for less than 2 months so I'm rapidly coming up to speed. Would this issue be better resolved by changing the order of how the renderer memory is cleaned up?

**Comment by TokisanGames** [comment]
I experience this problem in 4.1.3, and 2d3b2abcf6edcc9b02a9b1894ba43f79845ad2f4 using Terrain3D. Ever since @tcoxon added occlusion culling to Terrain3D. You can test it with this build of our plugin:
https://github.com/TokisanGames/Terrain3D/actions/runs/6774897812

1. Run the demo.
2. Click Terrain3D.
3. Click Bake Occluder3D on the top menu, and click OK. Wait 30s.
4. Save and restart the project

Godot crashes on `singleton->free_polygon(polygon_id);` 
```
Unhandled exception thrown: read access violation.
**RendererCanvasRender::singleton** was 0x1CDD576B050.

Callstack
>	[Inline Frame] godot.windows.editor.x86_64.exe!RendererCanvasRender::Polygon::{dtor}() Line 154	C++
 	godot.windows.editor.x86_64.exe!RendererCanvasRender::Item::CommandPolygon::`scalar deleting destructor'(unsigned int)	C++
 	[Inline Frame] godot.windows.editor.x86_64.exe!memdelete(RendererCanvasRender::Item::Command * p_class) Line 109	C++
 	godot.windows.editor.x86_64.exe!RendererCanvasRender::Item::clear() Line 416	C++
 	godot.windows.editor.x86_64.exe!RendererCanvasRender::Item::~Item() Line 470	C++
 	[External Code]	
 	godot.windows.editor.x86_64.exe!RID_Alloc<RendererCanvasCull::Item,1>::~RID_Alloc<RendererCanvasCull::Item,1>() Line 341	C++
 	godot.windows.editor.x86_64.exe!RendererCanvasCull::~RendererCanvasCull() Line 2163	C++
 	[External Code]	
 	[Inline Frame] godot.windows.editor.x86_64.exe!memdelete(RenderingServer * p_class) Line 109	C++
 	[Inline Frame] godot.windows.editor.x86_64.exe!finalize_display() Line 314	C++
 	godot.windows.editor.x86_64.exe!Main::cleanup(bool p_force) Line 3632	C++
 	godot.windows.editor.x86_64.exe!widechar_main(int argc, wchar_t * * argv) Line 184	C++
 	godot.windows.editor.x86_64.exe!_main() Line 206	C++
 	godot.windows.editor.x86_64.exe!main(int argc, char * * argv) Line 218	C++
 	[External Code]	

```

The fix above from @raulsntos works fine.
